### DISCUSSION TITLE
Guidance on how to develop/get started locally?

### DISCUSSION THREAD

**Discussion by biofreeze** [CardanoSolutions/kupo#154] [Q&A]
Hi new to Haskell/Cabal, so in terms of getting this running, i'm pretty lost. I've installed [https://www.haskell.org/ghcup](https://www.haskell.org/ghcup) via the website but i'm kinda lost what to do next.

I'm assuming I need to install a set of dependencies first, before I can develop locally? Can someone provide guidance on some commands I can run to get started? I would like to be able to make changes to the code for example, and then run it. And after i'm done with my changes, build it for Mac, Linux, etc.

I'm on a Mac (btw), which I think I saw is supported. Any help/guidance would be appreciated!

**Comment by KtorZ** [comment]
Hello, did you look at [CONTRIBUTING.md](https://github.com/CardanoSolutions/kupo/blob/master/CONTRIBUTING.md) already? 

I probably need to revise the dependency list as I think I forgot to do it for the last big Cardano upgrade. Although, the system pre requisite are identical to those of Ogmios which are detailed here: https://ogmios.dev/getting-started/building/#pre-requisites-server

On MacOS you might need some other extra flags in a local cabal.project configuration; but this depends on your architecture, shell and package manager. I can likely help with that as I am also mainly developing on Mac.

**Reply by biofreeze** [reply]
I did briefly skim over [CONTRIBUTING.md](https://github.com/CardanoSolutions/kupo/blob/master/CONTRIBUTING.md) page but tbh I got a little overwhelmed since I have never worked with Haskell or Cabal before lol. Regarding the dependencies listed on this page https://ogmios.dev/getting-started/building/#pre-requisites-server, I guess i'm not even sure how I would install those? Was honestly hoping there were just a list of commands to run and boom good to go 😆  Do I just find these libraries on Github and clone then `make install` them all or something? Again, i'll be brutally honest, I don't really have any experience using `make` either... so sorry on that front 🤣 

Once the prerequisites are installed, how would I say run Kupo as if I had the built version? (i.e - Start Kupo with the parameters like `./kupo --prune-utxo --ogmios-host wss://ogmios.us1.demeter.run --ogmios-port 443 --workdir ./db`) This repo is giving me a great opportunity to try to learn more about Haskell, so I wanted to explore and play around with the code, and then see if I could figure out how to rebuild it so I could run the optimized built version. (if that makes sense).

And my MacBook Pro is an M1 Max (2021) running Sonoma 14.2 🙂  (Not sure if that information helps)

**Reply by KtorZ** [reply]
Most of the system dependencies above you can install with homebrew. I suggest using your favorite search engine to see the exact packages you need to install. 

Those that aren't on homebrew have instructions following them (like the blst lib). 

Once the system deps are there, it's a very standard Haskell project, so: 

cabal update 
cabal build 

are enough :)

**Reply by biofreeze** [reply]
Ok here's what I've got so far. I was able to install the following packages via Brew:

libffi - https://formulae.brew.sh/formula/libffi
gmp - https://formulae.brew.sh/formula/gmp
pcre - https://formulae.brew.sh/formula/pcre
llvm - https://formulae.brew.sh/formula/llvm
pkgconf - https://formulae.brew.sh/formula/pkgconf
zlib - https://formulae.brew.sh/formula/zlib
libsodium - https://formulae.brew.sh/formula/libsodium
autoconf - https://formulae.brew.sh/formula/autoconf
libtool - https://formulae.brew.sh/formula/libtool
libssl via openssl - https://formulae.brew.sh/formula/openssl@3

I may need guidance on both of these packages: `libnuma-dev` and `libsystemd-dev` (systemd seems to be only for Linux?)

I did the following commands to install `A custom revision of bitcoin-core’s secp256k1, with Schnorr signature support enabled`:
```
git clone https://github.com/bitcoin-core/secp256k1.git
cd secp256k1
git reset --hard ac83be33d0956faf6b7f61a60ab524ef7d6a473a
./autogen.sh
./configure --prefix=/usr --enable-module-schnorrsig --enable-experimental
make
make check
sudo make install
```

The `make check` command resulted in a successful PASS:
```log
secp256k1 % make check
/Library/Developer/CommandLineTools/usr/bin/make  check-am
/Library/Developer/CommandLineTools/usr/bin/make  check-TESTS
PASS: tests
PASS: exhaustive_tests
============================================================================
Testsuite summary for libsecp256k1 0.1.0-pre
============================================================================
# TOTAL: 2
# PASS:  2
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

The last step there resulted in the followed error however:
```
secp256k1 % sudo make install
/Library/Developer/CommandLineTools/usr/bin/make  install-am
 mkdir -p '/usr/lib'
 /bin/sh ./libtool   --mode=install /usr/bin/install -c   libsecp256k1.la '/usr/lib'
libtool: install: /usr/bin/install -c .libs/libsecp256k1.0.dylib /usr/lib/libsecp256k1.0.dylib
install: /usr/lib/libsecp256k1.0.dylib: Operation not permitted
make[2]: *** [install-libLTLIBRARIES] Error 71
make[1]: *** [install-am] Error 2
make: *** [install] Error 2
```

**Reply by biofreeze** [reply]
Also when I run `cabal build` I get the following error
```
Warning: Requested index-state 2023-12-06T23:59:59Z is newer than
'hackage.haskell.org'! Falling back to older state (2023-12-06T23:57:21Z).
Warning: Requested index-state 2023-12-06T23:59:59Z is newer than
'cardano-haskell-packages'! Falling back to older state
(2023-12-06T19:40:56Z).
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: direct-sqlite-2.3.29 (user goal)
[__1] trying: base-4.18.1.0/installed-4.18.1.0 (dependency of direct-sqlite)
[__2] trying: kupo-0 (user goal)
[__3] trying: ouroboros-consensus-protocol-0.6.0.0 (dependency of kupo)
[__4] trying: cardano-crypto-tests-2.1.2.0 (dependency of
ouroboros-consensus-protocol)
[__5] next goal: cardano-crypto-praos (dependency of cardano-crypto-tests)
[__5] rejecting: cardano-crypto-praos-2.1.2.0, cardano-crypto-praos-2.1.1.2
(conflict: pkg-config package libsodium-any, not found in the pkg-config
database)
[__5] rejecting: cardano-crypto-praos-2.1.1.1 (conflict:
base==4.18.1.0/installed-4.18.1.0, cardano-crypto-praos => base>=4.14 &&
<4.17)
[__5] skipping: cardano-crypto-praos-2.1.1.0, cardano-crypto-praos-2.1.0.0,
cardano-crypto-praos-2.0.0.1, cardano-crypto-praos-2.0.0.0.1,
cardano-crypto-praos-2.0.0 (has the same characteristics that caused the
previous version to fail: excludes 'base' version 4.18.1.0)
[__5] fail (backjumping, conflict set: base, cardano-crypto-praos,
cardano-crypto-tests)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: base, ouroboros-consensus-protocol,
cardano-crypto-praos, cardano-crypto-tests, kupo, direct-sqlite
Try running with --minimize-conflict-set to improve the error message.
```

My `ghc --version` is the following:
```
ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.6.3
```

It look's like 2 packages are conflicting over different ghc versions or something? 🤔

**Reply by KtorZ** [reply]
Cabal build won't work until you have all the system dependencies since the constraint solver in cabal will discard versions that require those system dependencies, irremediably resulting in an invalid build plan. 

For libsodium, the issue seem to be a problem of permission. You don't have permissions to copy a file under /usr/lib. I suggest to configure the installation to a difference location that is owned by you and not root (eg /usr/local/lib). Usually you can do that via specific environment variables specified in the installation instructions or the make file directly.

**Reply by biofreeze** [reply]
Found somewhere that said to add the following to my `~/.bashrc`:
```
export LD_LIBRARY_PATH="/usr/local/lib:$LD_LIBRARY_PATH"
export PKG_CONFIG_PATH="/usr/local/lib/pkgconfig:/opt/homebrew/opt/openssl@3/lib/pkgconfig:$PKG_CONFIG_PATH"

#For compilers to find llvm you may need to set:
export LDFLAGS="-L/opt/homebrew/opt/llvm/lib -L/opt/homebrew/opt/openssl@3/lib"
export CPPFLAGS="-I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/openssl@3/include"
PATH="/opt/homebrew/opt/libtool/libexec/gnubin:/opt/homebrew/opt/openssl@3/lib/pkgconfig:/opt/homebrew:/opt/homebrew/bin:/opt/homebrew/opt/llvm/bin:$PATH"

export PATH="/opt/homebrew/opt/llvm/bin:$PATH"
```

And then do `source ~/.bashrc`.

But when I do `cabal build` I keep getting this error:
```
Failed to build cardano-crypto-class-2.1.4.0.
Build log (
/Users/biofreeze/.cache/cabal/logs/ghc-9.6.3/crdn-crypt-clss-2.1.4.0-24bd341a.log
):
Configuring library for cardano-crypto-class-2.1.4.0..
Preprocessing library for cardano-crypto-class-2.1.4.0..
running dist/build/Cardano/Crypto/Libsodium/Constants_hsc_make failed (exit code -6)
rsp file was: "dist/build/Cardano/Crypto/Libsodium/hsc2hscall3457-3.rsp"
output file:"dist/build/Cardano/Crypto/Libsodium/Constants.hs"
command was: dist/build/Cardano/Crypto/Libsodium/Constants_hsc_make  >dist/build/Cardano/Crypto/Libsodium/Constants.hs
error: dyld[3464]: Library not loaded: /usr/lib/libsecp256k1.0.dylib
  Referenced from: <78FD0050-8ACE-38FF-85E2-482E6731357D> /Users/biofreeze/Documents/kupo-repo/dist-newstyle/tmp/src-3374/cardano-crypto-class-2.1.4.0/dist/build/Cardano/Crypto/Libsodium/Constants_hsc_make
  Reason: tried: '/usr/lib/libsecp256k1.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/libsecp256k1.0.dylib' (no such file), '/usr/lib/libsecp256k1.0.dylib' (no such file, not in dyld cache)

Error: cabal: Failed to build cardano-crypto-class-2.1.4.0 (which is required
by test:unit from kupo-0 and exe:kupo from kupo-0). See the build log above
for details.
```

It's still using looking in `/usr/lib/` for some reason 😞
